chalk-rust-ir/src/lib.rs

```diff
@@ -422,6 +422,23 @@ pub struct AssociatedTyValueBound<I: Interner> {
     pub ty: Ty<I>,
 }
 
+/// Represents the bounds for an `impl Trait` type.
+///
+/// ```ignore
+/// opaque type T: A + B = HiddenTy;
+/// ```
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Fold)]
+pub struct ImplTraitDatum<I: Interner> {
+    /// The placeholder `!T` that corresponds to the opaque type `T`.
+    pub impl_trait_id: ImplTraitId<I>,
+
+    /// Trait bounds for the opaque type.
+    pub bounds: Vec<QuantifiedWhereClause<I>>,
```

Comment:
So, in the example of `T: A + B` above, what would the value of these where-clauses be? In particular, how is the `T` represented? My suggestion would be that we want to use a "binder" here for the 'self-type', kind of like we do for `dyn Write` or whatever. This is because sometimes we'll want to instantiate the where-clauses with the placeholder `!T` and sometimes (perhaps in follow-up PRs...) we'll want to instantiate those same clauses with the hidden type that is being inferred, etc.

Reply:
Oh, also, I think we really want something like:

```rust
 pub struct ImplTraitDatum<I: Interner> { 
     pub impl_trait_id: ImplTraitId<I>, 
     pub bound: Binders<ImplTraitDatumBound<I>>,
}
```

where `ImplTraitDatumBound` contains the where clauses + the hidden type. The reason is that both of them may reference the parameter kinds:

```rust
type Foo<T> = impl Iterator<Item = T> // hidden type = Vec<T>
```